apps/designer/app/routes/rest/project.$projectId.tsx

```diff
@@ -0,0 +1,97 @@
+import { LoaderFunction, json } from "@remix-run/node";
+import env from "~/env.server";
+import { db } from "@webstudio-is/project/server";
+import { sentryException } from "~/shared/sentry";
+import { generateCssText } from "~/shared/css-utils";
+import type { Props, Tree } from "@webstudio-is/react-sdk";
+import type { Page } from "@webstudio-is/project";
+
+type PagesDetails = {
+  [path: string]: {
+    page: Page;
+    tree: Tree | null;
+    props: Props[];
+    css: string;
+  };
+};
+export const loader: LoaderFunction = async ({ request, params }) => {
+  try {
+    const projectId = params.projectId ?? undefined;
+    const url = new URL(request.url);
+    const pagesDetails: PagesDetails = {};
+
+    if (projectId === undefined) {
+      throw json("Required project id", { status: 400 });
+    }
+
+    const prodBuild = await db.build.loadByProjectId(projectId, "prod");
+    if (prodBuild === undefined) {
+      throw json(
+        `Project ${projectId} not found or not published yet. Please contact us to get help.`,
+        { status: 500 }
+      );
+    }
+    const {
+      pages: { homePage, pages },
+    } = prodBuild;
+    const breakpoints = await db.breakpoints.load(prodBuild.id);
+    const tree = await db.tree.loadById(homePage.treeId);
+    if (tree === null) {
+      throw json(
+        `Tree ${homePage.treeId} structure not found. Please contact us to get help.`,
+        {
+          status: 500,
+        }
+      );
+    }
+    pagesDetails["/"] = {
+      page: homePage,
+      tree,
+      props: await db.props.loadByTreeId(homePage.treeId),
+      css: await generateCssText({
```

Comment:
hmm, now every rest call would be generating the CSS text and also delivering CSS as part of the tree, kindof duplicated data quite a bit, I think we should probably only do that as an option?

Reply:
I also just checked you are loading the build in generateCssText function and you do the same here  ... you will produce here lots of unnecessary queries to the database

Reply:
I think that you broke the fundamental idea behind the API: allowing user to decide how they render anything on the canvas.

API was supposed to deliver all the data so that user can compile whatever they need: any CSS they need, any components they need.

In the attempt to optimize the data fetching you changed the very premise of the api.

Reply:
Now even if we add an option for CSS either being a string or part of the tree, you don't have all the data user needs to build their CSS: e.g. breakpoints.

Reply:
I think you need to step back and only deliver data, without compiling anything. Compilation step should be still up to the consumer of the data and we need to take great care that those functions are easy to use and fast.